crate/client/src/datasets.rs

```diff
@@ -0,0 +1,100 @@
+use cosmian_findex_structs::{EncryptedEntries, Uuids};
+use tracing::{instrument, trace};
+use uuid::Uuid;
+
+use crate::{
+    error::{result::FindexClientResult, FindexClientError},
+    handle_error,
+    rest_client::SuccessResponse,
+    FindexRestClient,
+};
+
+impl FindexRestClient {
+    #[instrument(ret(Display), err, skip_all)]
+    pub async fn add_entries(
+        // todo(manu): revisit function names (prefix with dataset_, findex_, permissions)
+        &self,
+        index_id: &Uuid,
+        encrypted_entries: &EncryptedEntries,
+    ) -> FindexClientResult<SuccessResponse> {
+        let endpoint = format!("/datasets/{index_id}/add_entries");
+        let server_url = format!("{}{endpoint}", self.client.server_url);
+        trace!("POST: {server_url}");
+        let encrypted_entries = encrypted_entries.serialize()?;
+        let response = self
+            .client
+            .client
+            .post(server_url)
+            .body(encrypted_entries)
+            .send()
+            .await?;
+
+        trace!("Response: {response:?}");
+        let status_code = response.status();
+        if status_code.is_success() {
+            return Ok(response.json::<SuccessResponse>().await?);
+        }
+
+        // process error
+        let p = handle_error(&endpoint, response).await?;
+        Err(FindexClientError::RequestFailed(p))
```

Comment:
Avoid early returns when they are not necessary, they are surprising and thus require more time to parse.

```suggestion
        if status_code.is_success() {
            Ok(response.json::<SuccessResponse>().await?);
        } else {
            let p = handle_error(&endpoint, response).await?;
            Err(FindexClientError::RequestFailed(p))
        }
```

Reply:
done